### PR TITLE
feat(metadata): strengthen service-app mapping consistency, retry and…

### LINKED ISSUE #3354: [FEATURE] Improve service-app mapping consistency, retry, and deduplication / 加强 service-app mapping 的一致性、重试与去重
上下文请参考 https://github.com/apache/dubbo-go/issues/3188#issuecomment-4582464870

### Background

The `#2534` refactor kept service-app mapping as a core part of the application-level metadata path. Providers publish interface-to-application mappings, and consumers use those mappings to discover applications. This issue only focuses on application-level service-app mapping reliability.

Related code:
- `metadata/mapping/metadata/service_name_mapping.go`
- `metadata/report/nacos/report.go`
- `metadata/report/zookeeper/report.go`
- `metadata/report/etcd/report.go`

### Current Problems

The current mapping write path is relatively thin.

Specific problems:
- `retryTimes = 10` is hardcoded and has no sleep or backoff.
- Nacos, Etcd, and Zookeeper mostly use read-modify-write with comma-separated content.
- Concurrent writes may overwrite each other.
- Etcd and Zookeeper use `strings.Contains` to detect existing app names, which can produce substring false positives.
- Some implementations do not actually remove mapping listeners.

### Suggestions

- Replace hardcoded retry loops with configurable retry times, retry interval, and optional backoff.
- Use CAS or version-aware updates to avoid concurrent write overwrites.
- Parse app names as a set and use exact deduplication.
- Complete listener removal where supported, or document unsupported behavior explicitly.
- Add tests for concurrent mapping writes, duplicate app names, and listener removal.

### 背景

`#2534` 重构后，service-app mapping 仍是应用级 metadata 主链路的核心部分。provider 发布接口到应用名的映射，consumer 通过该映射发现应用。本 issue 只关注应用级 service-app mapping 的可靠性。

相关代码：
- `metadata/mapping/metadata/service_name_mapping.go`
- `metadata/report/nacos/report.go`
- `metadata/report/zookeeper/report.go`
- `metadata/report/etcd/report.go`

### 当前问题

当前 mapping 写入链路比较薄。

具体问题：
- `retryTimes = 10` 是硬编码，并且没有 sleep 或 backoff。
- Nacos、Etcd、Zookeeper 基本使用读旧值、拼逗号、写回的方式。
- 并发写入时可能互相覆盖。
- Etcd 和 Zookeeper 使用 `strings.Contains` 判断 app 是否存在，可能出现子串误判。
- 部分实现没有真正移除 mapping listener。

### 建议

- 用可配置的 retry times、retry interval 和可选 backoff 替代硬编码重试循环。
- 使用 CAS 或版本感知更新，避免并发写覆盖。
- 将 app names 解析为 set，并使用精确去重。
- 在支持的实现中补齐 listener remove；不支持时明确文档化。
- 增加并发 mapping 写入、重复 app name、listener remove 的测试。

### PR DESCRIPTION
**Pull request by NeverENG**
### Description
Fixes #3354

## What this PR does

Fixes #3354. Hardens application-level **service-app mapping** (`interface -> app names`)
registration so it is correct under concurrent providers, and gives it a proper retry policy.
完善应用级 service-app mapping 的写入一致性、重试与去重。

## Background

The mapping value is a comma-separated set of application names stored under a single
interface key, shared by all providers of that interface. Registration is therefore a
read-modify-write, and the previous implementations had several reliability gaps.

### 1. Optimistic concurrency across all backends (no more lost updates)
Concurrent appends no longer clobber each other:
- **etcd**: `Get`+`Put` → `GetValAndRev` + `UpdateWithRev` (CAS on `ModRevision`), `Create` for first write.
- **zookeeper**: keeps versioned `SetContent`, now surfaces version conflicts instead of swallowing them.
- **nacos**: adds `CasMd5` optimistic lock.

Each backend wraps its native conflict (`ErrCompareFail` / `ErrBadVersion` / `ErrNodeExists` /
nacos publish failure) into a shared `report.ErrMappingCASConflict` sentinel via `%w`.

### 2. Graded retry (was: fixed loop, no backoff)
`registerWithRetry` retries **only** CAS conflicts (`errors.Is`) with exponential backoff + jitter,
and returns permanent errors (network/auth) immediately instead of burning the whole retry budget.
原来任何错误都空转重试 10 次且无 sleep,现在按错误类型分级重试。

### 3. Extract shared logic + fix two hidden bugs
- `report.MergeServiceAppMapping`: whole-element dedup. Fixes the `strings.Contains` substring
  false positive (registering `order` was wrongly treated as present when `order-service` existed)
  and the leading-comma bug (`"" + "," + app` → `",app"`).
- `report.DecodeServiceAppNames`: parse into a set, skipping empty elements.

### 4. Listener cleanup
- **zookeeper**: implemented removal via `CacheListener.RemoveKeyListeners` (was a silent `return nil`
  that leaked listeners).
- **etcd**: documents the mapping listener as unsupported instead of silently succeeding.

### 5. Tests
- Unit tests for the merge/decode helpers (incl. the substring and empty-value regressions).
- A concurrency test that **reproduces the lost-update bug** with the naive read-modify-write and
  **proves CAS preserves every writer** (200 writers / 20 concurrent readers). Passes under `-race`.

## Known limitation (documented in code)

Nacos `CasMd5` is an optimistic **UPDATE** and cannot guard the first **INSERT** (Nacos has no
create-if-absent primitive), so the initial concurrent registration of a brand-new interface can
still race. **etcd and zookeeper are not affected.** Left as a documented limitation; can be revisited
if Nacos exposes a SETNX-style primitive.

## Test

go test -race ./metadata/report/... ./metadata/mapping/...

### Checklist
- [x] I confirm the target branch is `develop`
- [x] Code has passed local testing
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Alanxtl** [comment]
先看一下应该没和 #3371 重复吧

File: metadata/report/nacos/report.go [hunk 1]
```diff
@@ -165,30 +162,42 @@ func (n *nacosMetadataReport) RegisterServiceAppMapping(key string, group string
...
+			// Nacos surfaces a CAS rejection and a transport error the same way, so they
+			// cannot be told apart here. Treat the failure as a retriable conflict rather
+			// than risk dropping a real concurrent update; the underlying error is preserved
+			// for diagnosis.
+			return fmt.Errorf("publish mapping %s (%v): %w", key, err, report.ErrMappingCASConflict)
+		}
+		return err
+	}
+	return nil
 }
```
Comment (Alanxtl) [inline metadata/report/nacos/report.go:212]: Nacos 读失败会被当成空旧值继续写，可能覆盖已有 mapping。 

这里 `oldVal, _ := n.getConfig(...)` 仍然吞掉错误。若 Nacos 读因为网络/权限/服务端异常失败，但 key 实际已有 `appA,appB`，这里会按 `oldVal == ""` 生成 `Content: appC` 且不带 `CasMd5`，最后可能无条件覆盖旧集合。这个 bug 在 base 已存在，但这个 PR 的核心就是“mapping consistency”，所以这里必须顺手修掉：`getConfig` error 应该返回或进入可控重试，不能继续 publish。


另外我们要把md5算法换掉[#3370](https://github.com/apache/dubbo-go/pull/3370)
casmd5会不会受到影响

File: metadata/report/zookeeper/listener.go [hunk 2]
```diff
@@ -116,6 +116,15 @@ func (l *CacheListener) RemoveListener(key string, listener mapping.MappingListe
...
 }
 
+// RemoveKeyListeners drops all listeners registered for key so its mapping change events stop
+// being dispatched. The dispatcher goroutine is shared by the whole mapping group and is kept
+// alive (other keys still need it); it is released when the report is closed. The key's
+// underlying ZooKeeper watch is not unregistered here, as ZkEventListener exposes no per-path
+// unlisten, so the server may keep sending now-ignored events for the key.
+func (l *CacheListener) RemoveKeyListeners(key string) {
+	l.keyListeners.Delete(key)
+}
```
Comment (Alanxtl) [inline metadata/report/zookeeper/listener.go:126]: Zookeeper listener 路径没有复用新的 decode helper，事件里仍可能带空 app。

129行func (l *CacheListener) DataChange(event remoting.Event) bool {

仍然直接 strings.Split 后全量 set.Add(e)。如果已有 legacy 值是 ,app、app,,other 这类格式，GetServiceAppMapping 会过滤空元素，但 ZK change event 会把 "" 发给 consumer。建议这里也改成 report.DecodeServiceAppNames(event.Content)，并补 listener event 测试。

File: metadata/report/nacos/report.go (hunk 1 above)
Comment (NeverENG) [reply metadata/report/nacos/report.go:212]: 问题已解决
换算法应该不影响，#3370 是 app+revision -> MetaDataInfo 的映射。#3373（本PR）是 service-app-mapping，interface 到 app 到映射，这里的 md5 是 nacos 为实现 CAS 乐观锁的指纹，修改 #3370 对本 pr 无影响

**Comment by Alanxtl** [comment]
pls update to latest develop branch to fix ci fail
然后修一下sonarcloud的ci fail

**Comment by Alanxtl** [comment]
这个 PR 和 #3371 同时改 metadata/report/nacos、etcd、zookeeper，最终合入前一定要 rebase 到最新 develop，确认 MetadataReport 新接口和 mapping CAS 改动都还在。

**Review by Alanxtl — APPROVED** [review]
lgtm

#3360、#3362 先合，低耦合。
#3367 作为并发安全底座。
#3369 合入，建立 registryId/report/cache 作用域。
#3370 rebase 到 #3367 + #3369 之后，特别检查 revision 计算不要绕开锁。
#3371 再合，接受 MetadataReport 接口扩展，并补确认 Snapshot() 与 #3367 锁语义一致。
#3373 最后 rebase，因为它和 #3371 同改 report backends；语义不重复，但文件冲突概率高。

**Review by AlexStocks — APPROVED** [review]
